File: sdk/android/tracelet-sdk/src/main/kotlin/com/ikolvi/tracelet/sdk/algorithm/RTree.kt
package com.ikolvi.tracelet.sdk.algorithm

import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min

/**
 * Axis-aligned bounding box for R-tree nodes.
 */
internal class RTreeBBox(
    var minLat: Double,
    var minLng: Double,
    var maxLat: Double,
    var maxLng: Double,
) {
    companion object {
        fun fromPoint(lat: Double, lng: Double) = RTreeBBox(lat, lng, lat, lng)
    }

    /** Expand this bbox to contain [other]. */
    fun expand(other: RTreeBBox) {
        if (other.minLat < minLat) minLat = other.minLat
        if (other.minLng < minLng) minLng = other.minLng
        if (other.maxLat > maxLat) maxLat = other.maxLat
        if (other.maxLng > maxLng) maxLng = other.maxLng
    }

    /** Whether this bbox intersects [other]. */
    fun intersects(other: RTreeBBox): Boolean {
        return minLat <= other.maxLat &&
                maxLat >= other.minLat &&
                minLng <= other.maxLng &&
                maxLng >= other.minLng
    }

    /** Area of the bbox (degree² — only used for ordering, not real area). */
    val area: Double get() = (maxLat - minLat) * (maxLng - minLng)

    /** Area increase needed to encompass [other]. */
    fun enlargement(other: RTreeBBox): Double {
        val newMinLat = min(minLat, other.minLat)
        val newMinLng = min(minLng, other.minLng)
        val newMaxLat = max(maxLat, other.maxLat)
        val newMaxLng = max(maxLng, other.maxLng)
        return (newMaxLat - newMinLat) * (newMaxLng - newMinLng) - area
    }
}

/**
 * A data entry stored in the R-tree leaf.
 */
internal class RTreeEntry<T>(
    val lat: Double,
    val lng: Double,
    val radius: Double,
    val data: T,
) {
    val bbox: RTreeBBox by lazy {
        val latOffset = radius / 111320.0
        val lngOffset = radius / (111320.0 * cos(lat * PI / 180.0))
        RTreeBBox(
            lat - latOffset,
            lng - lngOffset,
            lat + latOffset,
            lng + lngOffset,
        )
    }
}

/**
 * Internal R-tree node.
 */
internal class RTreeNode<T>(val isLeaf: Boolean) {
    val children = mutableListOf<RTreeNode<T>>()
    val entries = mutableListOf<RTreeEntry<T>>()
    var bbox = RTreeBBox(
        Double.MAX_VALUE,
        Double.MAX_VALUE,
        -Double.MAX_VALUE,
        -Double.MAX_VALUE,
    )

    fun recalculateBBox() {
        bbox = RTreeBBox(
            Double.MAX_VALUE,
            Double.MAX_VALUE,
            -Double.MAX_VALUE,
            -Double.MAX_VALUE,
        )
        if (isLeaf) {
            for (e in entries) bbox.expand(e.bbox)
        } else {
            for (c in children) bbox.expand(c.bbox)
        }
    }

    val count: Int get() = if (isLeaf) entries.size else children.size
}

/**
 * R-tree spatial index for efficient geofence proximity queries.
 *
 * Supports O(log n) circle and bounding-box queries, enabling 10,000+
 * geofences with sub-millisecond lookup times.
 *
 * Uses a classic R-tree with configurable branching factor [maxEntries].
 * On each location update, call [queryCircle] to find geofences within
 * the proximity radius instead of iterating all geofences.
 */
class RTree<T>(val maxEntries: Int = 8) {
    private val minEntries = (maxEntries * 0.4).toInt().coerceAtLeast(1)

    private var root: RTreeNode<T>? = null
    private var _size = 0

    /** Number of entries in the tree. */
    val size: Int get() = _size

    /** Whether the tree is empty. */
    val isEmpty: Boolean get() = _size == 0

    /**
     * Insert a geofence into the tree.
     *
     * [lat], [lng] is the center point, [radius] is in meters, [data] is
     * the associated geofence object.
     */
    fun insert(lat: Double, lng: Double, radius: Double, data: T) {
        val entry = RTreeEntry(lat, lng, radius, data)

        if (root == null) {
            root = RTreeNode(isLeaf = true)
            root!!.entries.add(entry)
            root!!.recalculateBBox()
            _size++
            return
        }

        val path = mutableListOf<RTreeNode<T>>()
        val leaf = chooseLeaf(root!!, entry.bbox, path)
        leaf.entries.add(entry)
        leaf.recalculateBBox()
        _size++

        if (leaf.entries.size > maxEntries) {
            splitLeafAndPropagate(leaf, path)
        } else {
            for (node in path) node.recalculateBBox()
        }
    }

    /**
     * Remove the first entry matching [data] (by equality).
     * Returns true if found and removed.
     */
    fun remove(data: T): Boolean {
        if (root == null) return false
        val removed = removeEntry(root!!, data)
        if (removed) {
            _size--
            if (_size == 0) {
                root = null
            } else if (!root!!.isLeaf && root!!.children.size == 1) {
                root = root!!.children.first()
            }
        }
        return removed
    }

    /**
     * Find all entries within [radiusMeters] of ([lat], [lng]).
     *
     * Uses the bounding box of the search circle for R-tree traversal,
     * then applies exact haversine distance as a post-filter.
     */
    fun queryCircle(lat: Double, lng: Double, radiusMeters: Double): List<T> {
        if (root == null) return emptyList()

        val latOffset = radiusMeters / 111320.0
        val lngOffset = radiusMeters / (111320.0 * cos(lat * PI / 180.0))
        val searchBBox = RTreeBBox(
            lat - latOffset,
            lng - lngOffset,
            lat + latOffset,
            lng + lngOffset,
        )

        val candidates = mutableListOf<T>()
        queryBBoxRecursive(root!!, searchBBox) { entry ->
            val dist = GeoUtils.haversine(lat, lng, entry.lat, entry.lng)
            if (dist <= radiusMeters + entry.radius) {
                candidates.add(entry.data)
            }
        }
        return candidates
    }

    /**
     * Find all entries whose bounding boxes intersect the given bbox.
     */
    fun queryBBox(
        minLat: Double,
        minLng: Double,
        maxLat: Double,
        maxLng: Double,
    ): List<T> {
        if (root == null) return emptyList()

        val searchBBox = RTreeBBox(minLat, minLng, maxLat, maxLng)
        val results = mutableListOf<T>()
        queryBBoxRecursive(root!!, searchBBox) { entry ->
            results.add(entry.data)
        }
        return results
    }

    /** Clear all entries. */
    fun clear() {
        root = null
        _size = 0
    }

    // ── Internal methods ────────────────────────────────────────────────────

    private fun chooseLeaf(
        node: RTreeNode<T>,
        bbox: RTreeBBox,
        path: MutableList<RTreeNode<T>>,
    ): RTreeNode<T> {
        if (node.isLeaf) return node

        path.add(node)

        var best: RTreeNode<T>? = null
        var bestEnlargement = Double.MAX_VALUE
        var bestArea = Double.MAX_VALUE

        for (child in node.children) {
            val enlargement = child.bbox.enlargement(bbox)
            val area = child.bbox.area
            if (enlargement < bestEnlargement ||
                (enlargement == bestEnlargement && area < bestArea)
            ) {
                best = child
                bestEnlargement = enlargement
                bestArea = area
            }
        }

        return chooseLeaf(best!!, bbox, path)
    }

    private fun splitLeafAndPropagate(leaf: RTreeNode<T>, path: MutableList<RTreeNode<T>>) {
        val entries = ArrayList(leaf.entries)
        leaf.entries.clear()

        val sibling = RTreeNode<T>(isLeaf = true)
        pickSeedsAndDistributeEntries(entries, leaf, sibling)

        leaf.recalculateBBox()
        sibling.recalculateBBox()

        insertSiblingIntoParent(leaf, sibling, path)
    }

    private fun pickSeedsAndDistributeEntries(
        entries: List<RTreeEntry<T>>,
        node: RTreeNode<T>,
        sibling: RTreeNode<T>,
    ) {
        var seed1 = 0
        var seed2 = 1
        var maxWaste = -Double.MAX_VALUE

        for (i in entries.indices) {
            for (j in i + 1 until entries.size) {
                val combined = RTreeBBox(
                    min(entries[i].bbox.minLat, entries[j].bbox.minLat),
                    min(entries[i].bbox.minLng, entries[j].bbox.minLng),
                    max(entries[i].bbox.maxLat, entries[j].bbox.maxLat),
                    max(entries[i].bbox.maxLng, entries[j].bbox.maxLng),
                )
                val waste = combined.area - entries[i].bbox.area - entries[j].bbox.area
                if (waste > maxWaste) {
                    maxWaste = waste
                    seed1 = i
                    seed2 = j
                }
            }
        }

        node.entries.add(entries[seed1])
        sibling.entries.add(entries[seed2])

        for (i in entries.indices) {
            if (i == seed1 || i == seed2) continue
            node.recalculateBBox()
            sibling.recalculateBBox()
            val e1 = node.bbox.enlargement(entries[i].bbox)
            val e2 = sibling.bbox.enlargement(entries[i].bbox)
            if (e1 <= e2) {
                node.entries.add(entries[i])
            } else {
                sibling.entries.add(entries[i])
            }
        }
    }

    private fun insertSiblingIntoParent(
        node: RTreeNode<T>,
        sibling: RTreeNode<T>,
        path: MutableList<RTreeNode<T>>,
    ) {
        if (path.isEmpty()) {
            val newRoot = RTreeNode<T>(isLeaf = false)
            newRoot.children.add(node)
            newRoot.children.add(sibling)
            newRoot.recalculateBBox()
            root = newRoot
            return
        }

        val parent = path.removeAt(path.size - 1)
        parent.children.add(sibling)
        parent.recalculateBBox()

        if (parent.children.size > maxEntries) {
            splitInternalAndPropagate(parent, path)
        } else {
            for (ancestor in path) ancestor.recalculateBBox()
        }
    }

    private fun splitInternalAndPropagate(
        node: RTreeNode<T>,
        path: MutableList<RTreeNode<T>>,
    ) {
        val children = ArrayList(node.children)
        node.children.clear()

        val sibling = RTreeNode<T>(isLeaf = false)

        var seed1 = 0
        var seed2 = 1
        var maxWaste = -Double.MAX_VALUE
        for (i in children.indices) {
            for (j in i + 1 until children.size) {
                val combined = RTreeBBox(
                    min(children[i].bbox.minLat, children[j].bbox.minLat),
                    min(children[i].bbox.minLng, children[j].bbox.minLng),
                    max(children[i].bbox.maxLat, children[j].bbox.maxLat),
                    max(children[i].bbox.maxLng, children[j].bbox.maxLng),
                )
                val waste = combined.area - children[i].bbox.area - children[j].bbox.area
                if (waste > maxWaste) {
                    maxWaste = waste
                    seed1 = i
                    seed2 = j
                }
            }
        }

        node.children.add(children[seed1])
        sibling.children.add(children[seed2])

        for (i in children.indices) {
            if (i == seed1 || i == seed2) continue
            node.recalculateBBox()
            sibling.recalculateBBox()
            val e1 = node.bbox.enlargement(children[i].bbox)
            val e2 = sibling.bbox.enlargement(children[i].bbox)
            if (e1 <= e2) {
                node.children.add(children[i])
            } else {
                sibling.children.add(children[i])
            }
        }

        node.recalculateBBox()
        sibling.recalculateBBox()

        insertSiblingIntoParent(node, sibling, path)
    }

    private fun queryBBoxRecursive(
        node: RTreeNode<T>,
        searchBBox: RTreeBBox,
        callback: (RTreeEntry<T>) -> Unit,
    ) {
        if (!node.bbox.intersects(searchBBox)) return

        if (node.isLeaf) {
            for (entry in node.entries) {
                if (entry.bbox.intersects(searchBBox)) {
                    callback(entry)
                }
            }
        } else {
            for (child in node.children) {
                queryBBoxRecursive(child, searchBBox, callback)
            }
        }
    }

    private fun removeEntry(node: RTreeNode<T>, data: T): Boolean {
        if (node.isLeaf) {
            val idx = node.entries.indexOfFirst { it.data == data }
            if (idx >= 0) {
                node.entries.removeAt(idx)
                node.recalculateBBox()
                return true
            }
            return false
        }

        for (child in node.children) {
            if (removeEntry(child, data)) {
                if (child.count == 0) {
                    node.children.remove(child)
                }
                node.recalculateBBox()
                return true
            }
        }
        return false
    }
}

File: sdk/rust-core/src/spatial/rtree.rs
use std::f64;

#[derive(Clone, Debug)]
pub struct RTreeBBox {
    pub min_lat: f64,
    pub min_lng: f64,
    pub max_lat: f64,
    pub max_lng: f64,
}

impl RTreeBBox {
    pub fn new(min_lat: f64, min_lng: f64, max_lat: f64, max_lng: f64) -> Self {
        Self { min_lat, min_lng, max_lat, max_lng }
    }

    pub fn from_point(lat: f64, lng: f64) -> Self {
        Self::new(lat, lng, lat, lng)
    }

    pub fn expand(&mut self, other: &RTreeBBox) {
        if other.min_lat < self.min_lat { self.min_lat = other.min_lat; }
        if other.min_lng < self.min_lng { self.min_lng = other.min_lng; }
        if other.max_lat > self.max_lat { self.max_lat = other.max_lat; }
        if other.max_lng > self.max_lng { self.max_lng = other.max_lng; }
    }

    pub fn intersects(&self, other: &RTreeBBox) -> bool {
        self.min_lat <= other.max_lat &&
        self.max_lat >= other.min_lat &&
        self.min_lng <= other.max_lng &&
        self.max_lng >= other.min_lng
    }

    pub fn area(&self) -> f64 {
        (self.max_lat - self.min_lat) * (self.max_lng - self.min_lng)
    }

    pub fn enlargement(&self, other: &RTreeBBox) -> f64 {
        let new_min_lat = self.min_lat.min(other.min_lat);
        let new_min_lng = self.min_lng.min(other.min_lng);
        let new_max_lat = self.max_lat.max(other.max_lat);
        let new_max_lng = self.max_lng.max(other.max_lng);
        (new_max_lat - new_min_lat) * (new_max_lng - new_min_lng) - self.area()
    }
}

#[derive(Clone)]
pub struct RTreeEntry<T> {
    pub lat: f64,
    pub lng: f64,
    pub radius: f64,
    pub data: T,
    pub bbox: RTreeBBox,
}

impl<T> RTreeEntry<T> {
    pub fn new(lat: f64, lng: f64, radius: f64, data: T) -> Self {
        let lat_offset = radius / 111320.0;
        let lng_offset = radius / (111320.0 * (lat * f64::consts::PI / 180.0).cos());
        let bbox = RTreeBBox::new(
            lat - lat_offset,
            lng - lng_offset,
            lat + lat_offset,
            lng + lng_offset,
        );
        Self { lat, lng, radius, data, bbox }
    }
}

pub struct RTreeNode<T> {
    pub is_leaf: bool,
    pub children: Vec<Box<RTreeNode<T>>>,
    pub entries: Vec<RTreeEntry<T>>,
    pub bbox: RTreeBBox,
}

impl<T> RTreeNode<T> {
    pub fn new(is_leaf: bool) -> Self {
        Self {
            is_leaf,
            children: Vec::new(),
            entries: Vec::new(),
            bbox: RTreeBBox::new(f64::INFINITY, f64::INFINITY, f64::NEG_INFINITY, f64::NEG_INFINITY),
        }
    }

    pub fn recalculate_bbox(&mut self) {
        self.bbox = RTreeBBox::new(f64::INFINITY, f64::INFINITY, f64::NEG_INFINITY, f64::NEG_INFINITY);
        if self.is_leaf {
            for e in &self.entries {
                self.bbox.expand(&e.bbox);
            }
        } else {
            for c in &self.children {
                self.bbox.expand(&c.bbox);
            }
        }
    }

    pub fn count(&self) -> usize {
        if self.is_leaf {
            self.entries.len()
        } else {
            self.children.len()
        }
    }

    fn insert(&mut self, entry: RTreeEntry<T>, max_entries: usize) -> Option<Box<RTreeNode<T>>> {
        if self.is_leaf {
            self.entries.push(entry);
            self.recalculate_bbox();
            if self.entries.len() > max_entries {
                return Some(Box::new(self.split_leaf()));
            } else {
                return None;
            }
        } else {
            let mut best_idx = 0;
            let mut best_enlargement = f64::INFINITY;
            let mut best_area = f64::INFINITY;

            for (i, child) in self.children.iter().enumerate() {
                let enlargement = child.bbox.enlargement(&entry.bbox);
                let area = child.bbox.area();
                if enlargement < best_enlargement || (enlargement == best_enlargement && area < best_area) {
                    best_idx = i;
                    best_enlargement = enlargement;
                    best_area = area;
                }
            }

            if let Some(sibling) = self.children[best_idx].insert(entry, max_entries) {
                self.children.push(sibling);
                self.recalculate_bbox();
                if self.children.len() > max_entries {
                    return Some(Box::new(self.split_internal()));
                }
            } else {
                self.recalculate_bbox();
            }
            return None;
        }
    }

    fn split_leaf(&mut self) -> RTreeNode<T> {
        let entries = std::mem::take(&mut self.entries);
        let mut sibling = RTreeNode::new(true);

        let mut seed1 = 0;
        let mut seed2 = 1;
        let mut max_waste = f64::NEG_INFINITY;

        for i in 0..entries.len() {
            for j in (i + 1)..entries.len() {
                let combined = RTreeBBox::new(
                    entries[i].bbox.min_lat.min(entries[j].bbox.min_lat),
                    entries[i].bbox.min_lng.min(entries[j].bbox.min_lng),
                    entries[i].bbox.max_lat.max(entries[j].bbox.max_lat),
                    entries[i].bbox.max_lng.max(entries[j].bbox.max_lng),
                );
                let waste = combined.area() - entries[i].bbox.area() - entries[j].bbox.area();
                if waste > max_waste {
                    max_waste = waste;
                    seed1 = i;
                    seed2 = j;
                }
            }
        }

        let mut e1_opt = None;
        let mut e2_opt = None;
        let mut remaining = Vec::new();

        for (i, entry) in entries.into_iter().enumerate() {
            if i == seed1 {
                e1_opt = Some(entry);
            } else if i == seed2 {
                e2_opt = Some(entry);
            } else {
                remaining.push(entry);
            }
        }

        self.entries.push(e1_opt.unwrap());
        sibling.entries.push(e2_opt.unwrap());

        for entry in remaining {
            self.recalculate_bbox();
            sibling.recalculate_bbox();
            let e1 = self.bbox.enlargement(&entry.bbox);
            let e2 = sibling.bbox.enlargement(&entry.bbox);
            if e1 <= e2 {
                self.entries.push(entry);
            } else {
                sibling.entries.push(entry);
            }
        }

        self.recalculate_bbox();
        sibling.recalculate_bbox();
        sibling
    }

    fn split_internal(&mut self) -> RTreeNode<T> {
        let children = std::mem::take(&mut self.children);
        let mut sibling = RTreeNode::new(false);

        let mut seed1 = 0;
        let mut seed2 = 1;
        let mut max_waste = f64::NEG_INFINITY;

        for i in 0..children.len() {
            for j in (i + 1)..children.len() {
                let combined = RTreeBBox::new(
                    children[i].bbox.min_lat.min(children[j].bbox.min_lat),
                    children[i].bbox.min_lng.min(children[j].bbox.min_lng),
                    children[i].bbox.max_lat.max(children[j].bbox.max_lat),
                    children[i].bbox.max_lng.max(children[j].bbox.max_lng),
                );
                let waste = combined.area() - children[i].bbox.area() - children[j].bbox.area();
                if waste > max_waste {
                    max_waste = waste;
                    seed1 = i;
                    seed2 = j;
                }
            }
        }

        let mut child1 = None;
        let mut child2 = None;
        let mut remaining = Vec::new();

        for (i, child) in children.into_iter().enumerate() {
            if i == seed1 {
                child1 = Some(child);
            } else if i == seed2 {
                child2 = Some(child);
            } else {
                remaining.push(child);
            }
        }

        self.children.push(child1.unwrap());
        sibling.children.push(child2.unwrap());

        for child in remaining {
            self.recalculate_bbox();
            sibling.recalculate_bbox();
            let e1 = self.bbox.enlargement(&child.bbox);
            let e2 = sibling.bbox.enlargement(&child.bbox);
            if e1 <= e2 {
                self.children.push(child);
            } else {
                sibling.children.push(child);
            }
        }

        self.recalculate_bbox();
        sibling.recalculate_bbox();
        sibling
    }

    fn remove<F>(&mut self, predicate: &F) -> bool
    where F: Fn(&T) -> bool {
        if self.is_leaf {
            if let Some(idx) = self.entries.iter().position(|e| predicate(&e.data)) {
                self.entries.remove(idx);
                self.recalculate_bbox();
                return true;
            }
            return false;
        }

        let mut removed = false;
        let mut i = 0;
        while i < self.children.len() {
            if self.children[i].remove(predicate) {
                if self.children[i].count() == 0 {
                    self.children.remove(i);
                }
                self.recalculate_bbox();
                removed = true;
                break;
            } else {
                i += 1;
            }
        }
        removed
    }
}

pub struct RTree<T> {
    max_entries: usize,
    root: Option<Box<RTreeNode<T>>>,
    size: usize,
}

impl<T> RTree<T> {
    pub fn new(max_entries: usize) -> Self {
        Self { max_entries, root: None, size: 0 }
    }

    pub fn size(&self) -> usize {
        self.size
    }

    pub fn is_empty(&self) -> bool {
        self.size == 0
    }

    pub fn insert(&mut self, lat: f64, lng: f64, radius: f64, data: T) {
        let entry = RTreeEntry::new(lat, lng, radius, data);
        if let Some(ref mut root) = self.root {
            if let Some(sibling) = root.insert(entry, self.max_entries) {
                let mut new_root = RTreeNode::new(false);
                let old_root = self.root.take().unwrap();
                new_root.children.push(old_root);
                new_root.children.push(sibling);
                new_root.recalculate_bbox();
                self.root = Some(Box::new(new_root));
            }
        } else {
            let mut root = RTreeNode::new(true);
            root.entries.push(entry);
            root.recalculate_bbox();
            self.root = Some(Box::new(root));
        }
        self.size += 1;
    }

    pub fn remove<F>(&mut self, predicate: F) -> bool
    where F: Fn(&T) -> bool {
        if let Some(ref mut root) = self.root {
            if root.remove(&predicate) {
                self.size -= 1;
                if self.size == 0 {
                    self.root = None;
                } else if !root.is_leaf && root.children.len() == 1 {
                    self.root = Some(root.children.remove(0));
                }
                return true;
            }
        }
        false
    }

    pub fn query_circle<'a>(&'a self, lat: f64, lng: f64, radius_meters: f64) -> Vec<&'a T> {
        let mut results = Vec::new();
        if let Some(ref root) = self.root {
            let lat_offset = radius_meters / 111320.0;
            let lng_offset = radius_meters / (111320.0 * (lat * f64::consts::PI / 180.0).cos());
            let search_bbox = RTreeBBox::new(
                lat - lat_offset,
                lng - lng_offset,
                lat + lat_offset,
                lng + lng_offset,
            );

            self.query_bbox_recursive(root, &search_bbox, &mut |entry| {
                let dist = crate::algorithms::geo_utils::haversine(lat, lng, entry.lat, entry.lng);
                if dist <= radius_meters + entry.radius {
                    results.push(&entry.data);
                }
            });
        }
        results
    }

    fn query_bbox_recursive<'a, F>(&'a self, node: &'a RTreeNode<T>, search_bbox: &RTreeBBox, callback: &mut F)
    where F: FnMut(&'a RTreeEntry<T>) {
        if !node.bbox.intersects(search_bbox) { return; }

        if node.is_leaf {
            for entry in &node.entries {
                if entry.bbox.intersects(search_bbox) {
                    callback(entry);
                }
            }
        } else {
            for child in &node.children {
                self.query_bbox_recursive(child, search_bbox, callback);
            }
        }
    }

    pub fn clear(&mut self) {
        self.root = None;
        self.size = 0;
    }
}
